pulsar-client-go/pulsar/schema.go

```diff
@@ -0,0 +1,494 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package pulsar
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"reflect"
+	"unsafe"
+
+	log "github.com/apache/pulsar/pulsar-client-go/logutil"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/linkedin/goavro"
+)
+
+type SchemaType int
+
+const (
+	NONE         SchemaType = iota //No schema defined
+	STRING                         //Simple String encoding with UTF-8
+	JSON                           //JSON object encoding and validation
+	PROTOBUF                       //Protobuf message encoding and decoding
+	AVRO                           //Serialize and deserialize via Avro
+	BOOLEAN                        //
+	INT8                           //A 8-byte integer.
+	INT16                          //A 16-byte integer.
+	INT32                          //A 32-byte integer.
+	INT64                          //A 64-byte integer.
+	FLOAT                          //A float number.
+	DOUBLE                         //A double number
+	_                              //
+	_                              //
+	_                              //
+	KEY_VALUE                      //A Schema that contains Key Schema and Value Schema.
+	BYTES        = -1              //A bytes array.
+	AUTO         = -2              //
+	AUTO_CONSUME = -3              //Auto Consume Type.
+	AUTO_PUBLISH = -4              // Auto Publish Type.
+)
+
+// Encapsulates data around the schema definition
+type SchemaInfo struct {
+	Name   string
+	Schema string
+	Type   SchemaType
+}
+
+func NewSchemaInfo(name string, schema string, schemaType SchemaType) *SchemaInfo {
+	si := &SchemaInfo{
+		Name:   name,
+		Schema: schema,
+		Type:   schemaType,
+	}
+	return si
+}
+
+type Schema interface {
+	Serialize(v interface{}) ([]byte, error)
+	UnSerialize(data []byte, v interface{}) error
+	Validate(message []byte) error
+	GetSchemaInfo() *SchemaInfo
+}
+
+type SchemaDefinition struct {
+	SchemaDef *goavro.Codec
+}
+
+func NewSchemaDefinition(schema *goavro.Codec) *SchemaDefinition {
+	schemaDef := &SchemaDefinition{
+		SchemaDef: schema,
+	}
+	return schemaDef
+}
+
+// initCodec returns a Codec used to translate between a byte slice of either
+// binary or textual Avro data and native Go data.
+func initCodec(codec string) (*goavro.Codec, error) {
+	return goavro.NewCodec(codec)
+}
+
+type JsonSchema struct {
+	SchemaDefinition
+	SchemaInfo
+}
+
+func NewJsonSchema(codec string) *JsonSchema {
```

Comment:
`codec` is a confused noun here. we should call it `jsonAvroSchemaDefinition` or `jsonAvroSchemaDef` - it is a json string of the avro schema definition